### PR TITLE
Elasticsearch index backup

### PR DESCRIPTION
**Pull request by djsd123**
The script backs up and uploads the specified indices.  At the moment I have the named the elasticsearch repo to kibana-dashboards but we could name it something more generic in case we wanted to use this script to backup other indexes on other hosts in future. The fact that a specified index may not exist will not cause the command to fail and it should just continue.

### CODE REVIEW

**Comment by tijmenb** [comment]
@MatMoore can we use this ^ for search indices?

File: modules/govuk_elasticsearch/spec/classes/govuk_elasticsearch_spec.rb [hunk 1]
```diff
@@ -23,11 +23,11 @@
 
   describe '#manage_repo' do
     let(:params) {{
-      :version => '1.7.0',
+      :version => '1.4.3',
```
Comment (rboulton) [inline modules/govuk_elasticsearch/spec/classes/govuk_elasticsearch_spec.rb]: Why is this needed?  Will this downgrade our cluster to using elasticsearch 1.4.3?

**Comment by MatMoore** [comment]
@tijmenb in theory yes, this PR is doing is the same thing as our rummager code.

For more info about what we did, see the [rummager implentation](https://github.com/alphagov/rummager/tree/master/lib/snapshot) and the [previous PR against puppet](https://github.com/alphagov/govuk-puppet/pull/4196)

The complications we ran into were around pruning old snapshots, and ensuring that multiple snapshot operations (create and delete) don't run concurrently.

If you don't prune the snapshots, then I think the time taken degrades over time. Some discussion of this here: https://groups.google.com/forum/#!msg/elasticsearch/cXl9UrqgwBI/et56xPUqwGkJ

The things that will differ for rummager:

- There is currently nginx server between rummager and the elasticsearch cluster - I think this introduces some issues with request timeouts, since the api requests can take a long time to complete
- The code doesn't exist yet, but if we add tools for restoring snapshots, the name of the index matters even if its aliased

Will the configuration of the ES repository be handled by puppet as well?

File: modules/govuk_elasticsearch/templates/es-backup-s3.erb [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+# indices: option allows us to specify indexes in a comma separated list rather than snap-shooting everything. However we can set
+# the $ES_REPO variable to '_all' to snapshot all indices.
+#
+# ignore_unavailable: option when true will cause indices that do not exist to be ignored during snapshot
+#
+# include_global_state: option will prevent the cluster global state from being stored as part of the snapshot.
+# The entire snapshot will fail if one or more indices participating in the snapshot don’t have all primary shards available
+
+
+/usr/bin/envdir <%= @env_dir %>/env.d /usr/bin/curl --connect-timeout 10 -sS -XPUT "http://127.0.0.1:9200/_snapshot/"${ES_REPO}"?wait_for_completion=true&pretty" -d '{
```
Comment (MatMoore) [inline modules/govuk_elasticsearch/templates/es-backup-s3.erb]: Is this conflating the "create repository" call with the "create snapshot" call? I think this just creates the initial repository if it doesn't exist.

I expected the snapshot one to look like `/_snapshot/$ES_REPO/$SNAPSHOT_NAME`

https://www.elastic.co/guide/en/elasticsearch/reference/current/modules-snapshots.html#_snapshot

File: modules/govuk_elasticsearch/spec/classes/govuk_elasticsearch_spec.rb (hunk 1 above)
Comment (deanwilson) [inline modules/govuk_elasticsearch/spec/classes/govuk_elasticsearch_spec.rb]: These are test values for rspec and are only used when running spec tests, they are not visible, and will not be used, on actual puppet runs.

File: modules/govuk_elasticsearch/templates/es-backup-s3.erb (hunk 2 above)
Comment (djsd123) [inline modules/govuk_elasticsearch/templates/es-backup-s3.erb]: Hi Matt,
Yes I somehow left out the command to create the snapshot.  Thanks for the heads up.

I think we need to talk about the best way to implement some sort of housekeeping job to ensure we don't experience slow down when the repo gets too large.  I was told this was specifically for kibana dashboard indices but hoped it could be used more widely.

File: modules/govuk_elasticsearch/manifests/backup.pp [hunk 3]
```diff
@@ -0,0 +1,92 @@
+# == Class: govuk_elasticsearch::backup
+#
+# GOV.UK specific class to backup elasticsearch indexes.
+#
+# === Parameters
+#
+# [*aws_access_key_id*]
```
Comment (surminus) [inline modules/govuk_elasticsearch/manifests/backup.pp]: Empty description.

File: modules/govuk_elasticsearch/manifests/backup.pp [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    owner   => $user,
+    group   => $user,
+    mode    => '0550',
+    content => template('govuk_elasticsearch/es-backup-s3.erb'),
+  }
+
+  cron { 's3-elasticsearch-indexes':
+    command => '/usr/local/bin/es-backup-s3',
+    user    => $user,
+    hour    => 0,
```
Comment (surminus) [inline modules/govuk_elasticsearch/manifests/backup.pp]: Need to specify minute.

**Comment by surminus** [comment]
I think the code looks good.

File: modules/govuk_elasticsearch/manifests/backup.pp [hunk 5]
```diff
@@ -40,12 +40,17 @@
   $env_dir = '/etc/es_s3backup',
   $user = backup::client::govuk-backup,
   $es_repo = undef,
+  $enabled = fasle,
```
Comment (MatMoore) [inline modules/govuk_elasticsearch/manifests/backup.pp]: you have a typo here

File: modules/govuk_elasticsearch/templates/es-backup-s3.erb [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+  "bucket": "'"${S3_BUCKET}"'",
+  "region": "'"${AWS_REGION}"'",
+  "access_key": "'"${AWS_ACCESS_KEY_ID}"'",
+  "secret_key": "'"${AWS_SECRET_ACCESS_KEY}"'"
+  },
+  "compress": "true"
+  }'
+
+
+# indices: option allows us to specify indexes in a comma separated list rather than snap-shooting everything. However we can set
```
Comment (MatMoore) [inline modules/govuk_elasticsearch/templates/es-backup-s3.erb]: this should be "snapshotting" rather than "snap-shooting"

Comment (rjbate) [inline modules/govuk_elasticsearch/templates/es-backup-s3.erb]: "snap-shooting" could be the opposite of "snapshotting"

File: hieradata/integration.yaml [hunk 7]
```diff
@@ -39,6 +39,10 @@ govuk_cdnlogs::monitoring_enabled: false
 govuk_crawler::seed_enable: true
 
 govuk_elasticsearch::dump::run_es_dump_hour: '9'
+govuk_elasticsearch::backup::s3_bucket: 'kibana-dashboards'
```
Comment (MatMoore) [inline hieradata/integration.yaml]: on naming: maybe we could name these things after the cluster itself, rather than the index?

File: modules/govuk_elasticsearch/templates/es-backup-s3.erb [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+# ignore_unavailable: option when true will cause indices that do not exist to be ignored during snapshot
+#
+# include_global_state: option will prevent the cluster global state from being stored as part of the snapshot.
+# The entire snapshot will fail if one or more indices participating in the snapshot don’t have all primary shards available
+
+if [ $? -eq 0 ]; then
+
+  /usr/bin/envdir <%= @env_dir %>/env.d /usr/bin/curl --connect-timeout 10 -sS -XPUT "http://127.0.0.1:9200/_snapshot/"${ES_REPO}"/"${ES_SNAPSHOT}"?wait_for_completion=true&pretty" -d '{
+    "type": "s3",
+    "settings": {
```
Comment (MatMoore) [inline modules/govuk_elasticsearch/templates/es-backup-s3.erb]: I think you can take `type` and `settings` out of this request - the ES repo is configured in the previous one.

**Comment by MatMoore** [comment]
Will the cron job have a passive icinga alert associated with it?

**Comment by djsd123** [comment]
Yes, I can add the passive check.

File: modules/govuk_elasticsearch/templates/es-backup-s3.erb [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+  "secret_key": "'"${AWS_SECRET_ACCESS_KEY}"'"
+  },
+  "compress": "true"
+  }'
+
+
+# indices: option allows us to specify indexes in a comma separated list rather than snap-shooting everything. However we can set
+# the $ES_REPO variable to '_all' to snapshot all indices.
+#
+# ignore_unavailable: option when true will cause indices that do not exist to be ignored during snapshot
```
Comment (djsd123) [inline modules/govuk_elasticsearch/templates/es-backup-s3.erb]: Agreed.  When I added it I was unsure of how and where the script would be invoked.

File: modules/govuk_elasticsearch/templates/es-backup-s3.erb [hunk 10]
```diff
@@ -0,0 +1,71 @@
+#!/bin/bash
+
+set -e
+
+# Redirect stdout and stderr to syslog
+exec 1> >(/usr/bin/logger -s -t $(basename $0)) 2>&1
+
+# The default Icinga passive alert assumes that the script failed
+NAGIOS_CODE=2
+NAGIOS_MESSAGE="CRITICAL: Mongodb backup push to S3 failed"
```
Comment (alexmuller) [inline modules/govuk_elasticsearch/templates/es-backup-s3.erb]: There are 2 references to MongoDB in this file which should probably be Elasticsearch.

Comment (djsd123) [inline modules/govuk_elasticsearch/templates/es-backup-s3.erb]: Sorry. Sloppy copy&paste.  Fixed.

File: modules/govuk_elasticsearch/manifests/backup.pp [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+    file { 'es-backup-s3':
+      path    => '/usr/local/bin',
+      owner   => $user,
+      group   => $user,
+      mode    => '0550',
+      content => template('govuk_elasticsearch/es-backup-s3.erb'),
+    }
+
+  # Remove cron job if backup is not enabled
+    if $enabled {
```
Comment (surminus) [inline modules/govuk_elasticsearch/manifests/backup.pp]: Wrap this in the entire class

File: modules/govuk_elasticsearch/templates/es-backup-s3.erb [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+
+# Triggered whenever this script exits, successful or otherwise. The values
+# of CODE/MESSAGE will be taken from that point in time.
+function nagios_passive () {
+printf "<%= @ipaddress %>\t<%= @service_desc %>\t${NAGIOS_CODE}\t${NAGIOS_MESSAGE}\n" | /usr/sbin/send_nsca -H alert.cluster >/dev/null
+}
+
+trap nagios_passive EXIT
+
+/usr/bin/envdir <%= @env_dir %>/env.d  /usr/bin/curl --connect-timeout 10 -sS -XPUT "http://127.0.0.1:9200/_snapshot/"${ES_REPO}"?wait_for_completion=true&pretty" -d '{
```
Comment (surminus) [inline modules/govuk_elasticsearch/templates/es-backup-s3.erb]: Can we test if this would work just in a single string?

File: modules/govuk_elasticsearch/templates/es-backup-s3.erb [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+else
+STATUS=1
+fi
+
+if [ $STATUS -eq 0 ]; then
+NAGIOS_CODE=0
+NAGIOS_MESSAGE="OK: Elasticsearch backup push to S3 succeeded"
+fi
+
+exit $STATUS
```
Comment (surminus) [inline modules/govuk_elasticsearch/templates/es-backup-s3.erb]: Refactor status

**Comment by surminus** [comment]
Couple of comments but other than that it looks OK. Probably merge when ready and test in Integration.

**Comment by djsd123** [comment]
Retiring this one in favour of this https://github.com/alphagov/govuk-puppet/pull/4637
